DESIGN.md

```diff
@@ -0,0 +1,211 @@
+# ALPS
+Because every Go package requires a clever 4-letter acronym.
+
+## Structure
+Minimum Go version will be 1.16 because that's what's installed on my machine, and Go modules will be used for dependency 
+management. The overall architecture of the example program implementing the ALPS library will be 3 separate 
+binaries. It could be done in one binary, with the job worker also providing the HTTPS server, and providing the
+CLI commands by checking to see if the job worker is already running first before launching a new worker. However,
+I think that designing it with this structure provides for a clean, simple implementation, provides a separation 
+of concerns, and makes future extensibility easier. For example, it's likely that a feature such as scheduling 
+jobs which occur in the future or are recurring would be developed. In that case, it would be best to have the 
+job worker able to operate entirely independently of the HTTP service to isolate it from failures.
+
+- *pkg*
+    - *ipc*: types and methods required for inter-process communication.
+    - *log*: implement a common logging interface accross all components. This will not be developed from scratch,
+    just adapt the logger used in other projects
+    - *alps*: library code for creating and listening to Arbitrary Linux Process Service.
+- *cmd*
+    - *api*: HTTP API server using mTLS authentication with ACL authorization. Will communicate with the job 
+    worker via RPC. Authentication and authorization will be done via middleware. Any endpoint will return 
+    `{"status": 401, "error": "you are not authorized to perform that action"}` if the request does not have a recognized
+    user.
+        - **POST /start** starts a process and returns an ID to use for getting output and control
+            - Request: 
+            ```json
+            {"job": "ls -lh /var/log"}
+            ```
+            - Success response: 
+            ```json
+            {
+                "status": 200, 
+                "jobID": 123, 
+                "owner": "user@example.com",
+                "startTime": "2020-08-18T12:00:00Z" 
+            }
+            ``` 
+            - Error response: 
+            ```json
+            {"status": 500, "error": "some failure"}
+            ```
+        - **POST /stop/:jobID** stops the job with the given ID
```

Comment:
How do you intend to stop a job?

Reply:
I think calling os.Process.Kill on the process will be the simplest. If I wanted to implement time restrictions then I'd use the CommandContext to create the process with a timeout and hold onto the cancel function for early termination.